kernel/src/scan/data_skipping.rs

```diff
@@ -304,83 +181,100 @@ impl DataSkippingFilter {
-                    Expr::ge(max_col.clone(), lit_int.clone()),
-                ]),
-            ),
-            (
-                column.clone().ne(lit_int.clone()),
-                Expr::or_from([
-                    Expr::gt(min_col.clone(), lit_int.clone()),
-                    Expr::lt(max_col.clone(), lit_int.clone()),
-                ]),
-            ),
-            (
-                lit_int.clone().ne(column.clone()),
-                Expr::or_from([
-                    Expr::gt(min_col.clone(), lit_int.clone()),
-                    Expr::lt(max_col.clone(), lit_int.clone()),
-                ]),
-            ),
-        ];
+    /// Retrieves the null count of a column, if it exists.
+    fn get_nullcount_stat(&self, col: &ColumnName) -> Option<Expr> {
+        Some(joined_column_expr!("nullCount", col))
+    }
+
+    /// Retrieves the row count of a column (parquet footers always include this stat).
+    fn get_rowcount_stat(&self) -> Option<Expr> {
+        Some(column_expr!("numRecords"))
+    }
 
-        for (input, expected) in cases {
-            let rewritten = as_data_skipping_predicate(&input).unwrap();
-            assert_eq!(rewritten, expected)
+    fn eval_partial_cmp(
+        &self,
+        ord: Ordering,
+        col: Expr,
+        val: &Scalar,
+        inverted: bool,
+    ) -> Option<Expr> {
+        let op = match (ord, inverted) {
+            (Ordering::Less, false) => BinaryOperator::LessThan,
+            (Ordering::Less, true) => BinaryOperator::GreaterThanOrEqual,
+            (Ordering::Equal, false) => BinaryOperator::Equal,
+            (Ordering::Equal, true) => BinaryOperator::NotEqual,
+            (Ordering::Greater, false) => BinaryOperator::GreaterThan,
+            (Ordering::Greater, true) => BinaryOperator::LessThanOrEqual,
+        };
+        Some(Expr::binary(op, col, val.clone()))
+    }
+
+    fn eval_scalar(&self, val: &Scalar, inverted: bool) -> Option<Expr> {
+        PredicateEvaluatorDefaults::eval_scalar(val, inverted).map(Expr::literal)
+    }
+
+    fn eval_is_null(&self, col: &ColumnName, inverted: bool) -> Option<Expr> {
+        let safe_to_skip = match inverted {
+            true => self.get_rowcount_stat()?, // all-null
+            false => Expr::literal(0i64),      // no-null
+        };
+        Some(Expr::ne(self.get_nullcount_stat(col)?, safe_to_skip))
+    }
+
+    fn eval_binary_scalars(
+        &self,
+        op: BinaryOperator,
+        left: &Scalar,
+        right: &Scalar,
+        inverted: bool,
+    ) -> Option<Expr> {
+        PredicateEvaluatorDefaults::eval_binary_scalars(op, left, right, inverted)
+            .map(Expr::literal)
+    }
+
+    fn finish_eval_variadic(
+        &self,
+        mut op: VariadicOperator,
+        exprs: impl IntoIterator<Item = Option<Expr>>,
+        inverted: bool,
+    ) -> Option<Expr> {
+        if inverted {
+            op = op.invert();
         }
```

Comment:
could avoid the `mut` with:
```suggestion
        let op = if inverted {
            op.invert()
        } else {
            op
        }
```

Reply:
Is `mut` somehow bad? I was using `mut` there specifically to avoid the `else`...

Reply:
nope, it's fine. it's just a little less "functional" :)